src/rust/engine/remote_provider/remote_provider_traits/src/lib.rs

```diff
@@ -0,0 +1,135 @@
+  pub cas_address: String,
+  pub instance_name: Option<String>,
+  pub headers: BTreeMap<String, String>,
+  pub tls_config: grpc_util::tls::Config,
+  pub chunk_size_bytes: usize,
+  pub rpc_timeout: Duration,
+  pub rpc_retries: usize,
+  pub rpc_concurrency_limit: usize,
+  pub capabilities_cell_opt: Option<Arc<OnceCell<ServerCapabilities>>>,
+  pub batch_api_size_limit: usize,
+}
+
+#[derive(Clone)]
+pub struct RemoteCacheProviderOptions {
+  // TODO: this is currently framed for the REAPI provider, with some options used by others, would
+  // be good to generalise
+  // TODO: this is structurally very similar to `RemoteOptions`: maybe they should be the same? (see
+  // comment in `choose_action_cache_provider` too)
+  pub instance_name: Option<String>,
+  pub action_cache_address: String,
+  pub root_ca_certs: Option<Vec<u8>>,
+  pub mtls_data: Option<(Vec<u8>, Vec<u8>)>,
+  pub headers: BTreeMap<String, String>,
+  pub concurrency_limit: usize,
+  pub rpc_timeout: Duration,
+}
+
+#[async_trait]
+pub trait ByteStoreProvider: Sync + Send + 'static {
+  /// Store the bytes readable from `file` into the remote store
+  async fn store_file(&self, digest: Digest, file: File) -> Result<(), String>;
+
+  /// Store the bytes in `bytes` into the remote store, as an optimisation of `store_file` when the
+  /// bytes are already in memory
+  async fn store_bytes(&self, digest: Digest, bytes: Bytes) -> Result<(), String>;
+
+  /// Load the data stored (if any) in the remote store for `digest` into `destination`. Returns
+  /// true when found, false when not.
+  async fn load(
+    &self,
+    digest: Digest,
+    destination: &mut dyn LoadDestination,
+  ) -> Result<bool, String>;
+
+  /// Return any digests from `digests` that are not (currently) available in the remote store.
+  async fn list_missing_digests(
+    &self,
+    digests: &mut (dyn Iterator<Item = Digest> + Send),
+  ) -> Result<HashSet<Digest>, String>;
+}
+
+/// Places that write the result of a remote `load`
+#[async_trait]
+pub trait LoadDestination: AsyncWrite + Send + Sync + Unpin + 'static {
+  /// Clear out the writer and start again, if there's been previous contents written
+  async fn reset(&mut self) -> std::io::Result<()>;
+}
+
+#[async_trait]
+impl LoadDestination for tokio::fs::File {
+  async fn reset(&mut self) -> std::io::Result<()> {
+    self.rewind().await?;
+    self.set_len(0).await
+  }
+}
+
+#[async_trait]
+impl LoadDestination for Vec<u8> {
+  async fn reset(&mut self) -> std::io::Result<()> {
+    self.clear();
+    Ok(())
+  }
+}
+
+/// This `ActionCacheProvider` trait captures the operations required to be able to cache command
+/// executions remotely.
+#[async_trait]
+pub trait ActionCacheProvider: Sync + Send + 'static {
+  async fn update_action_result(
+    &self,
+    action_digest: Digest,
+    action_result: ActionResult,
+  ) -> Result<(), String>;
+
+  async fn get_action_result(
+    &self,
+    action_digest: Digest,
+    build_id: &str,
```

Comment:
This has changed from `context: &process_execution::Context` to just `build_id: &str`, because that's all that's used and it ensures that this code doesn't need to depend on `process_execution`.